{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fd1a3fe5-0415-4eda-b3b8-d34bfb3df8bd",
   "metadata": {},
   "source": [
    "### Tokenize Cancer Dependency Map data in order to send it into the Geneformer embedding extractor\n",
    "- create an Anndata object with `.X` as the raw gene expression information, and `.obs` containing patient ID, cell line, and disease (as well as disease state -- cancerous or not)\n",
    "- converts from gene name to ensembl ID\n",
    "\n",
    "Part 2:\n",
    "- obtain highly variable genes (top 500), and restrict the Anndata object to these 500 genes\n",
    "- tokenize it using the Geneformer tokenizer and save the resulting dataset to disk"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e340c813-1d0d-4aa2-99ec-53f0c6f6e771",
   "metadata": {},
   "source": [
    "### Data processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fbc17a24-606a-47dd-904a-abe548f946d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-23 23:19:09.517816: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "imports done\n"
     ]
    }
   ],
   "source": [
    "# Imports\n",
    "\n",
    "import sys\n",
    "from pathlib import Path\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "from scipy.sparse import csr_matrix\n",
    "# import scvelo as scv\n",
    "import h5py\n",
    "import os\n",
    "import scipy.sparse as sp\n",
    "from datasets import load_from_disk\n",
    "# import pooch\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import rcParams\n",
    "import h5py\n",
    "from tqdm import tqdm\n",
    "import pickle\n",
    "import torch\n",
    "\n",
    "#change to path to your Geneformer directory\n",
    "sys.path.append('/work/magroup/kaileyhu/Geneformer')\n",
    "from geneformer import EmbExtractor\n",
    "from geneformer import TranscriptomeTokenizer\n",
    "\n",
    "print(\"imports done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0c5af25c-2a72-4411-a672-2412f3fe46f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found device cuda:0\n"
     ]
    }
   ],
   "source": [
    "# Check torch device\n",
    "device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')\n",
    "print(f\"Found device {device}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ee9bf574-ecb9-483b-ad1b-266fb271f725",
   "metadata": {},
   "outputs": [],
   "source": [
    "#files\n",
    "metadata = pd.read_csv(\"/work/magroup/kaileyhu/datasets/depmap/metadata.csv\")\n",
    "omics_expr = pd.read_csv(\"/work/magroup/kaileyhu/datasets/depmap/OmicsExpressionProteinCodingGenesTPMLogp1.csv\")\n",
    "\n",
    "metadata.set_index('ModelID', inplace = True)\n",
    "omics_expr.set_index(\"Unnamed: 0\", inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "22fa29a8-a78e-4788-8ce4-41f7ed7d4354",
   "metadata": {},
   "outputs": [],
   "source": [
    "# protein coding genes\n",
    "\n",
    "adata = ad.AnnData(omics_expr)\n",
    "adata.obs_names = [str(i).split(\" \")[0] for i in omics_expr.index]\n",
    "adata.var_names = [str(i).split(\" \")[0] for i in omics_expr.columns]\n",
    "\n",
    "adata.obs[\"patient_id\"] = omics_expr.index\n",
    "adata.obs[\"cell_line\"] = metadata[\"OncotreeLineage\"]\n",
    "adata.obs[\"disease\"] = metadata[\"OncotreePrimaryDisease\"]\n",
    "\n",
    "for patient in adata.obs_names:\n",
    "    adata.obs.loc[patient, 'cell_line'] = metadata.loc[patient, \"StrippedCellLineName\"]\n",
    "    if (metadata.loc[patient, \"OncotreePrimaryDisease\"] != \"Non-Cancerous\"):\n",
    "        adata.obs.loc[patient, 'disease_state'] = \"Cancerous\"\n",
    "    else:\n",
    "        adata.obs.loc[patient, 'disease_state'] = \"Non-Cancerous\"\n",
    "    adata.obs.loc[patient, 'disease'] = metadata.loc[patient, \"OncotreePrimaryDisease\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4e6f4919-5e18-4f80-93db-616b1c92b6d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import ensembl_df of gene -> ensembl id matrix\n",
    "\n",
    "ensembl_path = \"/work/magroup/kaileyhu/Geneformer/geneformer/ensembl_mapping_dict_gc95M.pkl\"\n",
    "\n",
    "def invert_dict(dict_obj):\n",
    "    return {v: k for k, v in dict_obj.items()}\n",
    "\n",
    "with open(ensembl_path, \"rb\") as f:\n",
    "    id_gene_dict = pickle.load(f)\n",
    "    gene_id_dict = invert_dict(id_gene_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f1fa73a5-6081-4bbd-a839-d2d59a689b5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "lst = []\n",
    "genes = []\n",
    "\n",
    "for gene in adata.var_names:\n",
    "    gene2 = gene.split(\" \")[0]\n",
    "    if gene2 in id_gene_dict:\n",
    "        lst.append(id_gene_dict[gene2])\n",
    "        genes.append(gene2)\n",
    "    else:\n",
    "        lst.append(None)\n",
    "\n",
    "\n",
    "filtered_results = []\n",
    "\n",
    "res = []\n",
    "for val in lst:\n",
    "    if val is not None:\n",
    "        filtered_results.append(val)\n",
    "\n",
    "adata2 = adata[:,genes]\n",
    "adata2.var_names = filtered_results \n",
    "adata2.var['ensembl_id'] = filtered_results\n",
    "adata2.obs['n_counts'] = adata2.X.sum(axis=1)\n",
    "adata2.obs[\"cell_type\"] = adata.obs['cell_line']\n",
    "adata2.obs[\"disease\"] = adata.obs['disease']\n",
    "adata2.obs[\"disease_state\"] = adata.obs['disease_state']\n",
    "adata2.obs[\"batch\"] = 0\n",
    "adata2.obs[\"patient_id\"] = adata.obs[\"patient_id\"]\n",
    "adata2.var['gene_name'] = genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "af631844-adb6-4ade-b653-8b645fad7e2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1479 × 18858\n",
       "    obs: 'patient_id', 'cell_line', 'disease', 'disease_state', 'n_counts', 'cell_type', 'batch'\n",
       "    var: 'ensembl_id', 'gene_name'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "96bd1c34-41eb-4b4d-9085-3c9a6949e081",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata2.write_h5ad(\"/work/magroup/kaileyhu/datasets/depmap/processed/omics_expr.h5ad\",compression='gzip')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "854958c5-7fb9-430c-9b1c-14fa723cf9ed",
   "metadata": {},
   "source": [
    "### Obtain highly variable genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d870b4c8-ce27-4923-ae24-21156dff7c76",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(adata2, n_top_genes=500, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80c1ce20-424e-412b-919e-89a62dbe812a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_hvg = adata2[:, adata2.var[\"highly_variable\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc1cac84-de8c-4123-9360-6a52d2bf4de4",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_hvg.write_h5ad(\"/work/magroup/kaileyhu/datasets/depmap/processed/hvg/omics_expr_hvg_500.h5ad\",compression='gzip')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e4e3f35-fa13-4114-bcb7-36c73b92a4c3",
   "metadata": {},
   "source": [
    "### Tokenization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7c1ebbc-615d-4067-91ca-95d6ff78d2ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "tk = TranscriptomeTokenizer({\"cell_type\": \"cell_type\", \"disease\": \"disease\", \"disease_state\": \"disease_state\", \"patient_id\" : \"patient_id\"}, \n",
    "                            nproc=16,\n",
    "                            special_token = True, #true for 95M\n",
    "                            model_input_size=4096)  \n",
    "\n",
    "# 18130 total genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dea4eb9e-957b-48a4-adb3-38a5e19371a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "tk.tokenize_data('/work/magroup/kaileyhu/datasets/depmap/processed/hvg/', \n",
    "                 \"/work/magroup/kaileyhu/res/\", \n",
    "                 \"hvg_500_tokenized\", \n",
    "                 file_format=\"h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31a2fe65-f85c-41cd-a329-ee957751ef17",
   "metadata": {},
   "source": [
    "### Tokenization (non-HVG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e0d4ccb-227d-4397-8563-43b31a97b5c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "tk = TranscriptomeTokenizer({\"cell_type\": \"cell_type\", \"disease\": \"disease\", \"disease_state\": \"disease_state\", \"patient_id\" : \"patient_id\"}, \n",
    "                            nproc=16,\n",
    "                            special_token = True, #true for 95M\n",
    "                            model_input_size=4096)  \n",
    "\n",
    "# 18130 total genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbef8476-ffbf-4431-a3c0-881f936d3852",
   "metadata": {},
   "outputs": [],
   "source": [
    "tk.tokenize_data('/work/magroup/kaileyhu/datasets/depmap/processed/', \n",
    "                 \"/work/magroup/kaileyhu/res/\", \n",
    "                 \"all_genes_tokenized\", \n",
    "                 file_format=\"h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fe302dc-9f2d-455a-a1b6-f4e001140eac",
   "metadata": {},
   "source": [
    "### Embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f0177d2-4e48-40bf-aba5-c9334d3b4cdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.listdir(\"../../../Geneformer/fine_tuned_models/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37fd2a03-255e-4fce-81e4-1da349aebbce",
   "metadata": {},
   "outputs": [],
   "source": [
    "extractor_type = \"cancer\"\n",
    "path_to_Geneformer = \"../../../Geneformer\"\n",
    "\n",
    "if extractor_type == \"cancer\":\n",
    "    model = f\"{path_to_Geneformer}/gf-12L-95M-i4096_CLcancer\"\n",
    "    n_classes = 71\n",
    "elif extractor_type == \"cardio\":\n",
    "    model = f\"{path_to_Geneformer}/fine_tuned_models/gf-6L-30M-i2048_CellClassifier_cardiomyopathies_220224\"\n",
    "    n_classes = 3\n",
    "else: # pretrained\n",
    "    model = f\"{path_to_Geneformer}/gf-12L-30M-i2048\"\n",
    "    n_classes = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7952825c-65b3-4ba9-b3c5-6596942c10f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# initiate EmbExtractor\n",
    "print(f\"Starting emb extractor...\")\n",
    "embex = EmbExtractor(model_type=\"CellClassifier\",\n",
    "                     num_classes=n_classes,\n",
    "                     emb_layer=-1,\n",
    "                     emb_label=[\"cell_type\",\"disease\", \"disease_state\"],\n",
    "                     labels_to_plot=[\"disease_state\", \"disease\"],\n",
    "                     forward_batch_size=200,\n",
    "                     nproc=16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a1233ff-fdf7-4200-a47e-befe2dcab5fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# extracts embedding from input data\n",
    "print(f\"Extracting embeddings from {extractor_type}...\")\n",
    "embs = embex.extract_embs(model,\n",
    "                          \"/work/magroup/kaileyhu/res/hvg_500_tokenized.dataset\",\n",
    "                          \"/work/magroup/kaileyhu/res/\",\n",
    "                          f\"omics_expr_original_embs_{extractor_type}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d45cc0e-a60e-4348-9da1-8ec5d1448911",
   "metadata": {},
   "outputs": [],
   "source": [
    "embex.plot_embs(embs=embs, \n",
    "                plot_style=\"umap\",\n",
    "                output_directory=\"/work/magroup/kaileyhu/res/\",  \n",
    "                output_prefix=\"emb_plot\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0129821-aed8-4255-8c1f-dc15abadd722",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
